### PR TITLE
Minor: Introduce utils::hash for StructArray

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.
We will need this for https://github.com/apache/arrow-datafusion/pull/7893

### CODE REVIEW

File: datafusion/common/src/hash_utils.rs [hunk 1]
```diff
@@ -207,6 +208,37 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
     Ok(())
 }
 
+fn hash_struct_array(
+    array: &StructArray,
+
```
Comment (waynexia) [inline datafusion/common/src/hash_utils.rs:213]: nit:
```suggestion
```

File: datafusion/common/src/hash_utils.rs [hunk 2]
```diff
@@ -327,12 +359,16 @@ pub fn create_hashes<'a>(
                 array => hash_dictionary(array, random_state, hashes_buffer, rehash)?,
                 _ => unreachable!()
             }
+            DataType::Struct(_) => {
+                let array = as_struct_array(array)?;
+                hash_struct_array(array, random_state, hashes_buffer)?;
+            }
             DataType::List(_) => {
-                let array = as_list_array(array);
+                let array = as_list_array(array)?;
```
Comment (waynexia) [inline datafusion/common/src/hash_utils.rs:367]: Just a question: I'm a little confused about this change. Where changes the signature of `as_list_array`?

**Review by waynexia — APPROVED** [review]
Looks good to me, thanks @jayzhan211

File: datafusion/common/src/hash_utils.rs (hunk 2 above)
Comment (Weijun-H) [reply datafusion/common/src/hash_utils.rs:367]: There are two `as_list_array ` with different signatures from arrow-rs and datafusion.

Comment (waynexia) [reply datafusion/common/src/hash_utils.rs:367]: Thanks for your explanation!

File: datafusion/common/src/hash_utils.rs [hunk 3]
```diff
@@ -207,6 +208,37 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
...
+
+    // Skip null columns
+    let valid_indices: Vec<usize> = if let Some(nulls) = nulls {
+        nulls.valid_indices().collect()
+    } else {
+        (0..num_columns).collect()
+    };
+
+    for i in valid_indices {
+        let mut values_hashes = vec![0u64; array.len()];
```
Comment (Weijun-H) [inline datafusion/common/src/hash_utils.rs:229]: Why do you create a new vec for each iteration?

Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:229]: create each values_hashes for each `column`. And combine those hashed element in column to one

Comment (Dandandan) [reply datafusion/common/src/hash_utils.rs:228]: The `values_hashes` can be created upfront, outside of the loop.

Comment (Dandandan) [reply datafusion/common/src/hash_utils.rs:228]: Or we can actually just use `hashes_buffer` which already contains the zeroed items and remove this `values_hashes`.

File: datafusion/common/src/hash_utils.rs [hunk 4]
```diff
@@ -207,6 +208,36 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
...
+        let column_values = values[i].clone();
+        // create hashes for each column
+        create_hashes(&[column_values], random_state, &mut values_hashes)?;
+        let hash = &mut hashes_buffer[i];
```
Comment (Dandandan) [inline datafusion/common/src/hash_utils.rs:232]: This seems not correct, as `i` here to a column in the struct and `hashes_buffer` to a hash of one value?

Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:232]: @Dandandan. Not fully got your point. I do consider each column as one value (iterate each row in one column with get one final hash).

`i` is the column index in the struct array and also is the hashes_buffer index.
For example, column1, null, column3, column4. we iterate 0, 2, and 3. And hash each of them to hashes_buffer[0], hashes_buffer[2], and hashes_buffer[3] respectively. A column-wise hashes.

Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:232]: I think I mistake how the null buffer works in StructArray. It seems to be row-wise, but I consider it as column-wise...

Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:231]: It looks pretty better now :)

File: datafusion/common/src/hash_utils.rs [hunk 5]
```diff
@@ -207,6 +208,32 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
...
+    hashes_buffer: &mut [u64],
+) -> Result<()> {
+    let nulls = array.nulls();
+    let num_columns = array.num_columns();
+
+    // Skip null columns
+    let valid_indices: Vec<usize> = if let Some(nulls) = nulls {
+        nulls.valid_indices().collect()
+    } else {
+        (0..num_columns).collect()
```
Comment (Dandandan) [inline datafusion/common/src/hash_utils.rs:223]: This doesn't yet seem correct, `valid_indices` contains the rows that are valid, not the columns.

File: datafusion/common/src/hash_utils.rs [hunk 6]
```diff
@@ -207,6 +208,32 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
...
+    };
+
+    let mut values_hashes = vec![0u64; array.len()];
+    create_hashes(array.columns(), random_state, &mut values_hashes)?;
```
Comment (Dandandan) [inline datafusion/common/src/hash_utils.rs:227]: We should be able to use `hashes_buffer` here directly?

Comment (Dandandan) [reply datafusion/common/src/hash_utils.rs:227]: If we do that, we should be able to remove the code related to `valid_indices`, as this already takes care of nulls.

File: datafusion/common/src/hash_utils.rs [hunk 7]
```diff
@@ -327,12 +354,16 @@ pub fn create_hashes<'a>(
                 array => hash_dictionary(array, random_state, hashes_buffer, rehash)?,
                 _ => unreachable!()
             }
+            DataType::Struct(_) => {
+                let array = as_struct_array(array)?;
+                hash_struct_array(array, random_state, hashes_buffer)?;
```
Comment (Dandandan) [inline datafusion/common/src/hash_utils.rs:359]: This should work?
```suggestion
                create_hashes(array.columns(), random_state, &mut hashes_buffer)?;
```

File: datafusion/common/src/hash_utils.rs (hunk 5 above)
Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:223]: I hash each row now, so we iterate the valid row and hash each column at that row to one hash value.

File: datafusion/common/src/hash_utils.rs (hunk 6 above)
Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:227]: I dont think so.

values_hashes: [9258723240401068087, 9258723240401068087, 8502738074356456021, 8502738074356456021, 4222447303697976283, 9753707356376286577]
hashes_buffer: [9258723240401091360, 9258723240401091360, 0, 8502738074356479294, 0, 0]

We can see that `create_hashes` does not consider if the row is valid or not. Therefore, we need to iterate `valid_indices` once .

File: datafusion/common/src/hash_utils.rs (hunk 7 above)
Comment (jayzhan211) [reply datafusion/common/src/hash_utils.rs:359]: If we consider nulls, this doesn't work

**Comment by alamb** [comment]
@Dandandan  do you think this PR is ready to merge?

**Comment by jayzhan211** [comment]
@alamb @Dandandan Any update on this?

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
Thank you @jayzhan211 @waynexia  and @jayzhan211

**Comment by jayzhan211** [comment]
I found the current implementation failed this

```
    #[test]
    // Tests actual values of hashes, which are different if forcing collisions
    #[cfg(not(feature = "force_hash_collisions"))]
    fn create_hashes_for_struct_arrays_more_column_than_row() {
        let struct_array = StructArray::from(
            vec![
                (
                    Arc::new(Field::new("bool", DataType::Boolean, false)),
                    Arc::new(BooleanArray::from(vec![
                        false, false
                    ])) as ArrayRef,
                ),
                (
                    Arc::new(Field::new("i32-1", DataType::Int32, false)),
                    Arc::new(Int32Array::from(vec![10, 10])) as ArrayRef,
                ),
                (
                    Arc::new(Field::new("i32-2", DataType::Int32, false)),
                    Arc::new(Int32Array::from(vec![10, 10])) as ArrayRef,
                ),
                (
                    Arc::new(Field::new("i32-3", DataType::Int32, false)),
                    Arc::new(Int32Array::from(vec![10, 10])) as ArrayRef,
                ),
            ],
        );

        assert!(struct_array.is_valid(0));
        assert!(struct_array.is_valid(1));

        let array = Arc::new(struct_array) as ArrayRef;

        let random_state = RandomState::with_seeds(0, 0, 0, 0);
        let mut hashes = vec![0; array.len()];
        create_hashes(&[array], &random_state, &mut hashes).unwrap();
        assert_eq!(hashes[0], hashes[1]);
    }
```

Let me checkout the reason

**Comment by jayzhan211** [comment]
It seems `valid_indices` is column indices not row indices.

**Pull request merged**